exp/internal/controllers/machinepool_controller_phases.go

```diff
@@ -197,48 +197,52 @@ func (r *MachinePoolReconciler) reconcileBootstrap(ctx context.Context, cluster
 			return ctrl.Result{}, nil
 		}
 		bootstrapConfig = bootstrapReconcileResult.Result
-	}
 
-	// If the bootstrap data secret is populated, set ready and return.
-	if m.Spec.Template.Spec.Bootstrap.DataSecretName != nil {
-		m.Status.BootstrapReady = true
-		conditions.MarkTrue(m, clusterv1.BootstrapReadyCondition)
-		return ctrl.Result{}, nil
-	}
+		// If the bootstrap config is being deleted, return early.
+		if !bootstrapConfig.GetDeletionTimestamp().IsZero() {
+			return ctrl.Result{}, nil
+		}
 
-	// If the bootstrap config is being deleted, return early.
-	if !bootstrapConfig.GetDeletionTimestamp().IsZero() {
-		return ctrl.Result{}, nil
-	}
+		// Determine if the bootstrap provider is ready.
+		ready, err := external.IsReady(bootstrapConfig)
+		if err != nil {
+			return ctrl.Result{}, err
+		}
 
-	// Determine if the bootstrap provider is ready.
-	ready, err := external.IsReady(bootstrapConfig)
-	if err != nil {
-		return ctrl.Result{}, err
-	}
+		// Report a summary of current status of the bootstrap object defined for this machine pool.
+		conditions.SetMirror(m, clusterv1.BootstrapReadyCondition,
+			conditions.UnstructuredGetter(bootstrapConfig),
+			conditions.WithFallbackValue(ready, clusterv1.WaitingForDataSecretFallbackReason, clusterv1.ConditionSeverityInfo, ""),
+		)
 
-	// Report a summary of current status of the bootstrap object defined for this machine pool.
-	conditions.SetMirror(m, clusterv1.BootstrapReadyCondition,
-		conditions.UnstructuredGetter(bootstrapConfig),
-		conditions.WithFallbackValue(ready, clusterv1.WaitingForDataSecretFallbackReason, clusterv1.ConditionSeverityInfo, ""),
-	)
+		if !ready {
+			log.V(2).Info("Bootstrap provider is not ready, requeuing")
+			m.Status.BootstrapReady = ready
+			return ctrl.Result{RequeueAfter: externalReadyWait}, nil
```

Comment:
@CecileRobertMichon @Jont828 I think it's a bit unrelated but while we're here. I think this requeu is not required because we're adding a watch for BootstrapConfigs in reconcileExternal. This means we automatically get a reconcile if the BootstrapConfig changes.

Or maybe I'm missing something

Reply:
Hm but it's the same for InfrastructureReady. Maybe we can open an issue for a follow-up PR?

Reply:
@sbueringer My understanding was that we mainly wanted to break out of the reconcile loop without an error if we couldn't proceed in this case. Do you mean that we don't want to return a `ctrl.Result{}` here?

Reply:
I mean that we should return `return ctrl.Result{}, nil`

There is no reason to requeue if we get a reconcile triggered as soon as BootstrapConfig changes.

But let's keep this change out of this PR

Reply:
Opened an issue: https://github.com/kubernetes-sigs/cluster-api/issues/8735